src/Cardano/Transaction/Balancer/Collateral.purs

```diff
@@ -63,50 +63,89 @@ setTransactionCollateral
   -> Record (BalancerProtocolParameters r)
   -> Address
   -> Transaction
+  -> UtxoMap
   -> BalanceTxM Transaction
 setTransactionCollateral
   getWalletCollateralAff
   balancerConstraints
   pparams
   changeAddr
-  transaction = do
-  let
-    nonSpendableSet = balancerConstraints ^. _nonSpendableInputs
-    mbCollateralUtxos = balancerConstraints ^. _collateralUtxos
-  -- We must filter out UTxOs that are set as non-spendable in the balancer
-  -- constraints
-  let
-    isSpendable = not <<< flip Set.member nonSpendableSet
-    coinsPerUtxoByte = pparams.coinsPerUtxoByte
-  collateral <- case mbCollateralUtxos of
-    -- if no collateral utxos are specified, use the wallet, but filter
-    -- the unspendable ones
-    Nothing -> do
-      let isSpendableUtxo = isSpendable <<< _.input <<< unwrap
-      (walletCollateral :: Array TransactionUnspentOutput) <- ExceptT $ liftAff
-        $
-          note CouldNotGetCollateral <$>
+  transaction
+  spendableUtxos = do
+  collateral <-
+    case (unwrap balancerConstraints).collateralUtxos of
+      -- if no collateral utxos are specified, use the wallet, but filter
+      -- the unspendable ones
+      Nothing -> do
+        (walletCollateral :: Array TransactionUnspentOutput) <-
+          ExceptT $ liftAff $ note CouldNotGetCollateral <$>
             getWalletCollateralAff
-      let
-        { yes: spendableUtxos, no: filteredUtxos } = Array.partition
-          isSpendableUtxo
-          walletCollateral
-      when (not $ Array.null filteredUtxos) do
-        logWithLevel Warn $ pprintTagSet
-          "Some of the collateral UTxOs returned by the wallet were marked as non-spendable and ignored"
-          (pprintUtxoMap (TransactionUnspentOutputs.toUtxoMap filteredUtxos))
-      pure spendableUtxos
-    -- otherwise, get all the utxos, filter out unspendable, and select
-    -- collateral using internal algo, that is also used in KeyWallet
-    Just utxoMap -> do
-      let
-        maxCollateralInputs = UInt.toInt pparams.maxCollateralInputs
-        mbCollateral =
-          Array.fromFoldable <$>
-            selectCollateral coinsPerUtxoByte maxCollateralInputs utxoMap
-      liftEither $ note (InsufficientCollateralUtxos utxoMap) mbCollateral
+        let
+          { yes: spendableCollUtxos, no: filteredUtxos } = Array.partition
+            isSpendable
+            walletCollateral
+        when (not $ Array.null filteredUtxos) do
+          logWithLevel Warn $ pprintTagSet
+            "Some of the collateral UTxOs returned by the wallet were marked as non-spendable and ignored"
+            (pprintUtxoMap $ toUtxoMap filteredUtxos)
+        let
+          collVal =
+            foldMap (Val.fromValue <<< _.amount <<< unwrap <<< _.output <<< unwrap)
+              spendableCollUtxos
+          minRequiredCollateral = BigNum.toBigInt $ unwrap Collateral.minRequiredCollateral
```

Comment:
This is a constant of 5_000_000 lovelace, however, it is specified to be at least some percentage of tx fee based on point 4 mentioned in this section of this CIP https://github.com/cardano-foundation/CIPs/tree/master/CIP-0040#motivation-why-is-this-cip-necessary

But, I guess, it would have to be decided in multiple steps and it would be overly complex, since tx fee can be calculated once a TX is balanced....
Right?